From d52aaa5a5ebf281f96eac136ba68e63295c5bb8e Mon Sep 17 00:00:00 2001
From: usa <usa@b2dd03c8-39d4-4d8f-98ff-823fe69b080e>
Date: Wed, 14 Jan 2015 06:55:00 +0000
Subject: [PATCH] merge revision(s) 48534: [Backport #10412]

	* lib/resolv.rb: fall back if canonicalization fails.
	  Thanks Vit Ondruch for the patch!  [ruby-core:65836]

	* test/resolv/test_dns.rb: test for patch


git-svn-id: svn+ssh://ci.ruby-lang.org/ruby/branches/ruby_2_0_0@49245 b2dd03c8-39d4-4d8f-98ff-823fe69b080e
---
 ChangeLog               |  7 +++++++
 lib/resolv.rb           |  4 ++++
 test/resolv/test_dns.rb | 20 ++++++++++++++
 3 files changed, 32 insertions(+), 0 deletions(-)

diff --git a/ChangeLog b/ChangeLog
index 1aea7c0..8ea4238 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,10 @@
+Wed Jan 14 15:54:18 2015  Aaron Patterson <aaron@tenderlovemaking.com>
+
+	* lib/resolv.rb: fall back if canonicalization fails.
+	  Thanks Vit Ondruch for the patch!  [ruby-core:65836]
+
+	* test/resolv/test_dns.rb: test for patch
+
 Thu Nov 13 22:31:50 2014  CHIKANAGA Tomoyuki  <nagachika@ruby-lang.org>
 
 	* lib/rexml/document.rb: add REXML::Document#document.
diff --git a/lib/resolv.rb b/lib/resolv.rb
index 53bfd54..a302d8a 100644
--- a/lib/resolv.rb
+++ b/lib/resolv.rb
@@ -945,6 +945,10 @@ def generate_candidates(name)
             candidates = []
           end
           candidates.concat(@search.map {|domain| Name.new(name.to_a + domain)})
+          fname = Name.create("#{name}.")
+          if !candidates.include?(fname)
+            candidates << fname
+          end
         end
         return candidates
       end
diff --git a/test/resolv/test_dns.rb b/test/resolv/test_dns.rb
new file mode 100644
index 0000000..0ab5966
--- /dev/null
+++ b/test/resolv/test_dns.rb
@@ -0,0 +1,20 @@
+require 'test/unit'
+require 'resolv'
+
+class TestResolvDNS < Test::Unit::TestCase
+
+  # [ruby-core:65836]
+  def test_resolve_with_2_ndots
+    conf = Resolv::DNS::Config.new :nameserver => ['127.0.0.1'], :ndots => 2
+    assert conf.single?
+
+    candidates = []
+    conf.resolv('example.com') { |candidate, *args|
+      candidates << candidate
+      raise Resolv::DNS::Config::NXDomain
+    }
+    n = Resolv::DNS::Name.create 'example.com.'
+    assert_equal n, candidates.last
+  end
+
+end
